physionet-django/project/authorization/access.py

```diff
@@ -0,0 +1,123 @@
+from datetime import datetime
+
+from django.db.models import Q
+
+from events.models import Event, EventDataset
+from project.models import AccessPolicy, DUASignature, DataAccessRequest, PublishedProject
+from user.models import Training, TrainingType
+
+
+def get_public_projects_query():
+    return Q(access_policy=AccessPolicy.OPEN)
+
+
+def get_restricted_projects_query(user):
+    dua_signatures = DUASignature.objects.filter(user=user)
+    query = Q(access_policy=AccessPolicy.RESTRICTED) & Q(duasignature__in=dua_signatures)
+    return query
+
+
+def get_credentialed_projects_query(user):
+    dua_signatures = DUASignature.objects.filter(user=user)
+    completed_training = (
+        Training.objects.get_valid()
+        .filter(user=user)
+        .values_list("training_type")
+    )
+    not_completed_training = TrainingType.objects.exclude(pk__in=completed_training)
+    required_training_complete = ~Q(required_trainings__in=not_completed_training)
+
+    accepted_data_access_requests = DataAccessRequest.objects.filter(
+        requester=user, status=DataAccessRequest.ACCEPT_REQUEST_VALUE
+    )
+
+    contributor_review_with_access = Q(
+        access_policy=AccessPolicy.CONTRIBUTOR_REVIEW
+    ) & Q(data_access_requests__in=accepted_data_access_requests)
+
+    credentialed_with_dua_signed = Q(
+        access_policy=AccessPolicy.CREDENTIALED
+    ) & Q(duasignature__in=dua_signatures)
+
+    query = required_training_complete & (
+        contributor_review_with_access | credentialed_with_dua_signed
+    )
+    return query
+
+
+def get_projects_accessible_through_events(user):
+    events_all = Event.objects.filter(Q(host=user) | Q(participants__user=user))
+    active_events = set(events_all.filter(end_date__gte=datetime.now()))
+    accessible_datasets = EventDataset.objects.filter(event__in=active_events, is_active=True)
+    accessible_projects_ids = []
+    for event_dataset in accessible_datasets:
+        if event_dataset.has_access(user):
+            accessible_projects_ids.append(event_dataset.dataset.id)
+    query = Q(id__in=accessible_projects_ids)
+    return query
+
+
+def get_accessible_projects(user):
+    """
+    Returns query filter for published projects accessible by a specified user
+    """
+    query = Q(deprecated_files=False)
+
+    query &= get_public_projects_query()
+
+    if user.is_authenticated:
+        query |= get_restricted_projects_query(user)
+
+    if user.is_credentialed:
+        query |= get_credentialed_projects_query(user)
+
+    query |= get_projects_accessible_through_events(user)
+
+    return PublishedProject.objects.filter(query).distinct()
+
+
+def can_access_project(project, user):
```

Comment:
Thinking about the function names, why does `can_view_project_files` include "files", but `can_access_project` does not? Does it relate to the purpose of these functions?:

1) e.g. is `can_access_project` is actually determining whether or not someone can access the project description and files (e.g. for an unpublished project) while
2) `can_view_project_files` specifically relates to the files?

If so, makes sense! Just checking...

Reply:
From what i saw in the code before refactor we had

`is_authorized` ---> Checks if a user has access to project (user might still not have access to the files download/view)
`can_view_files`--->  checks `is_authorized`+ if the allow_file_downloads boolean is turned on for the project

So here also, my idea was has `can_access_project` checks if they have access to the project (files might not be accessible if `allow_file_downloads` is off) -->also  on the User interface, i think if this boolean is off, we dont display the files to the user on project page

Reply:
Sorry, if i made things further confusing or misunderstood. Happy to jump on a call to clarify things quickly

Reply:
I updated the doc string, to try to explain this better.